hw_iterators_generators.py

```diff
@@ -0,0 +1,132 @@
+from typing import Tuple, Generator
+import random
+
+
+class FastaRandomPrinter:
+    """Class object reads fasta file infinite,
+        path - string with path to file,
+        mutation_prob - float in range [0, 1] defining seq probability to mutate
+        float probability
+    """
+    def __init__(self, path, mutation_prob):
+        self.path = path
+        self.id = ''
+        self.seq = []
+        self.mutation_prob = mutation_prob
+        self.f = open(self.path, 'r', encoding='utf-8')
+        self.entry = ''
+
+    def get_mutated_seq(self, seq):
+        amino_acids = ("G", "P", "A", "V", "L", "I", "M",
+                       "C", "F", "Y", "W", "H", "K", "R",
+                       "Q", "N", "E", "D", "S", "T")
+        replacements_num = random.randint(1, len(seq) - 1)
+        positions = random.sample(tuple(range(0, len(seq) - 1)), k=replacements_num)
+        seq_list = [x if i not in positions
+                    else random.choice(amino_acids) for i, x in enumerate(seq)]
+        return ''.join(seq_list)
+
+    def __iter__(self):
+        return self
+
+    def __next__(self):
+        while True:
+            line = self.f.readline().rstrip('\n')
+            if line:
+                if line.startswith('>'):
+                    if self.id:
+                        if random.uniform(0, 1) < self.mutation_prob:
+                            self.entry = f"{self.id} {self.get_mutated_seq(''.join(self.seq))}"
+                        else:
+                            self.entry = f"{self.id} {''.join(self.seq)}"
+                    self.id = line
+                    self.seq = []
+                    return self.entry
+                else:
+                    self.seq.append(line)
+            else:
+                self.f.close()
+                self.f = open(self.path, 'r', encoding='utf-8')
+
+
+def fasta_reader(path_fasta) -> Generator[Tuple[str, str], str, str]:
+    """Generator function to read fasta files
+        Args:
+            path_fasta - the path to file in fasta format
+
+        Yields:
+             tuple with id: str, seq: str
+    """
+    with open(path_fasta, 'r', encoding='utf-8') as f:
+        id = ''
```

Comment:
`id` - встроенная функция, эту переменную лучше не перезаписывать
```suggestion
        id_ = ''
```